pyomo/contrib/doe/measurements.py

```diff
@@ -25,252 +25,270 @@
-
-        self.flatten_measure_name = flatten_names
-
-    def _generate_variance(self, flatten_measure_name, variance, name_and_index):
+        Used for generating string names with indexes. 
+
+        Parameter 
+        ---------
+        var_name: a ``list`` of var names 
+        extra_index: a ``list`` containing extra indexes except for time indexes 
+            if default (None), no extra indexes needed for all var in var_name
+            if it is a nested list, it is a ``list`` of ``list`` of ``list``, 
+            they are different multiple sets of indexes for different var_name
+            for e.g., extra_index[0] are all indexes for var_name[0], extra_index[0][0] are the first index for var_name[0]
+        time_index: a ``list`` containing time indexes
+            default choice is None, means this is a model parameter 
+            if it is an algebraic variable, time index should be set up to [0]
+            if it is a nested list, it is a ``list`` of ``lists``, they are different time set for different var in var_name
         """
-        Generate the variance dictionary
-        Parameters
-        ----------
-        flatten_measure_name: flattened measurement names. For e.g., flattenning {'C':{'CA': 10, 'CB': 1, 'CC': 2}} will be 'C_index_CA', ..., 'C_index_CC'.
-        variance:
-            a ``dict``, keys are measurement variable names, values are a dictionary, keys are its extra index name,
-            values are its variance as a scalar number.
-            For e.g., for the kinetics illustrative example, it should be {'C':{'CA': 10, 'CB': 1, 'CC': 2}}.
-            If given None, the default is {'C':{'CA': 1, 'CB': 1, 'CC': 1}}.
-            If there is no extra index, it is a dict, keys are measurement variable names, values are its variance as a scalar number.
-        name_and_index:
-            a dictionary, keys are measurement names, values are a list of extra indexes.
+
+        for i, n in enumerate(var_name):
+            name_data = str(n)
+
+            # first combine all indexes into a list 
+            all_index_list = [] # contains all index lists
+            if extra_index:
+                for index_list in extra_index[i]: 
+                    all_index_list.append(index_list)
+            if time_index:
+                all_index_list.append(time_index[i])
+
+            # all idnex list for one variable, such as ["CA", 10, 1]
+            all_index_for_var = list(itertools.product(*all_index_list))
+
+            for lst in all_index_for_var:
+                name1 = name_data+"["
+                for i, idx in enumerate(lst):
+                    name1 += str(idx)
+
+                    # if i is the last index, close the []. if not, add a "," for the next index. 
+                    if i==len(lst)-1:
+                        name1 += "]"
+                    else:
+                        name1 += ","
+
+                self.special_set.append(name1)
```

Comment:
Values and bounds may also need to be handled here. Again, changing from calling this function once to multiple times will require refactoring the logic.